### PR TITLE
Add try catch blocks to handle compile exceptions

### PR DESCRIPTION
**Pull request by Glitch18**

### CODE REVIEW

**Comment by JorikSchellekens** [comment]
Can you switch this to a draft?

**Comment by swapnilraj** [comment]
I am not familiar with the details of the parser and the analyzer but on first thought they probably throw some exceptions as well? You probably know more about this, if they do throw exceptions we should catch them as well.

File: libwarp/solidity_prepass/WarpVisitor.cpp [hunk 1]
```diff
@@ -498,7 +498,14 @@ void WarpVisitor::dynFuncArgsPass(const char* solFilepath) {
...
     this->m_compiler->analyze();
-    this->m_compiler->compile();
+
+    try{
+        this->m_compiler->compile();
+    } catch (boost::exception const& ex) {
+        auto error = this->m_compiler->errors().rbegin();
+        std::cout << "Compilation error : " << error->get()->what() << std::endl;
+        throw std::runtime_error("Compilation failed");
+    }
```
Comment (murcake) [inline libwarp/solidity_prepass/WarpVisitor.cpp:508]: Why not rethrow the exception?
Why are we only printing the last error?
Why are we not logging any info from `ex`?

**Review by murcake — COMMENTED** [review]
Is it a better idea to just wrap up the whole main in exception handlers and loggers? (perhaps extract the main itself into a separate function without hanlding)

**Comment by Glitch18** [comment]
> I am not familiar with the details of the parser and the analyzer but on first thought they probably throw some exceptions as well? You probably know more about this, if they do throw exceptions we should catch them as well.

Thanks for the suggestion @swapnilraj! Actually, `m_compiler->parse()` and `m_compiler()->analyze()` don't throw exceptions. They add their errors to the `ErrorList` of `m_compiler`. This list is checked when `m_compiler->compile()` is called and this is where exceptions are thrown. 

In fact, our calls to `parse()` and `analyze()` are redundant. `compile()` has an internal check where it calls both of these functions if they haven't been called before compilation. I'm thinking of replacing our explicit calls with these internal ones, in the case that they are not being called explicitly for a reason. This will clean up the code a little and put all three functions , `parse()`, `analyze()` and `compile()` in the same try catch block. Let me know what you think of this.

**Comment by Glitch18** [comment]
Thanks for the comments @murcake, I'll implement these and rebase my PR.

File: libwarp/solidity_prepass/WarpVisitor.cpp (hunk 1 above)
Comment (Glitch18) [reply libwarp/solidity_prepass/WarpVisitor.cpp]: > Why not rethrow the exception?

Thanks. This has been implemented.

> Why are we only printing the last error?

I've now used a for loop which iterates only over the error that were generated during the corresponding compile command.

> Why are we not logging any info from `ex`?

I had tried this, and the only info that I could get from `ex` was a message saying `Compile called with errors`. I felt our error reporting made this redundant and hence I'm not logging anything from `ex`. Is there something more that I can use from `ex`? Please let me know, thanks!

File: kudu.cpp [hunk 2]
```diff
@@ -51,6 +51,20 @@ int printUsageInfo(char* argv[]) {
...
 
+void printYulAST(std::string& main_contract, std::string& contractContents, 
+    const char* sol_filepath, bool print_ir) {
+    try {
+      WarpVisitor warpVisitor(main_contract, contractContents, sol_filepath,
+                             print_ir);
+      warpVisitor.yulPrepass().yulPass().generateYulAST();
+      std::cout << warpVisitor.m_yul_JSON_AST << std::endl;
+    } catch (std::exception const &exc) {
+      std::cerr << "ERROR: " << exc.what() << "\n";
```
Comment (murcake) [inline kudu.cpp:62]: Any reason not to use std::endl? It causes flushing and errors are useful to see as soon as they arise

File: libwarp/solidity_prepass/WarpVisitor.cpp (hunk 1 above)
Comment (murcake) [reply libwarp/solidity_prepass/WarpVisitor.cpp]: If you rethrow, it's okay not to log the exception itself

File: libwarp/solidity_prepass/WarpVisitor.cpp [hunk 3]
```diff
@@ -351,9 +360,18 @@ void WarpVisitor::refreshCompilerState(string filepath) {
...
+    try{
+        if(!this->m_compiler->compile()) throw std::runtime_error("Compilation failed");
+    } catch (std::exception const& ex) {
+        for(size_t i = start; i < this->m_compiler->errors().size(); i++) {
+            auto err = this->m_compiler->errors()[i];
+            if(err->errorSeverity(err->type()) != solidity::langutil::Error::Severity::Error) continue;
+            std::cout << "Compilation error : " << this->m_compiler->errors()[i].get()->what() << std::endl;
+        }
+        throw;
+    }
```
Comment (murcake) [inline libwarp/solidity_prepass/WarpVisitor.cpp:374]: There is some possibility for duplicate code extraction, but it's logically probably not a part of this PR

**Comment by Glitch18** [comment]
Examples:

Considering a faulty Solidity contract:
```
$:  cat ../test.sol
contract thisCon {
        bool true;
}
```

Current message upon compiling:
```
$: ./kudu --yul-json-ast ../test.sol thisCon
terminate called after throwing an instance of 'boost::wrapexcept<solidity::langutil::CompilerError>'
  what():  Called compile with errors.
[1]    4508 abort      ./kudu --yul-json-ast ../test.sol thisCon
```

Message with error reporting:
```
$: ./kudu --yul-json-ast ../test.sol thisCon
Compilation error : Expected identifier but got 'true'
ERROR: Compilation failed
```

**Review by murcake — APPROVED** [review]
Just one thing, could you please clang-format all the changes using the config in the repository root?

File: kudu.cpp (hunk 2 above)
Comment (Glitch18) [reply kudu.cpp:62]: No, you're right. I meant to use `std::endl`. Must have missed it here.

**Comment by Glitch18** [comment]
I've added the `std::endl`s and clang-formatted my changes.

**Review by JorikSchellekens — APPROVED** [review]

**Comment by JorikSchellekens** [comment]
Some pretty bad news, Greg's cleanup operation has torn through this pr

**Comment by Glitch18** [comment]
I see. No worries, I'll implement it again.

**Pull request merged**